### ISSUE TITLE
Timer function replays when singleton lock lost

### ISSUE THREAD

**Issue by mathewc** [Azure/azure-functions-host#2968]
As described [here](https://github.com/Azure/azure-webjobs-sdk-extensions/wiki/TimerTrigger#singleton-locks) timer trigger functions rely on a blob lease behind the scenes to ensure only a single instance of the timer is running across scaled out instances.

In normal processing, an instance is long lived, holds the lease and monitors the schedule, invoking the function at the right time. Behind the scenes the lease is periodically renewed to ensure no other host runs the timer.

However, in cases where the lease is lost during the invocation of the timer function, another host instance can start up, get the lease and start running the timer **while the other invocation is still completing**. A lease can be lost due to due to host level failure, resource exhaustion/thread starvation preventing the lease from being renewed, etc. Because of our host graceful shutdown, active function invocations aren't terminated, they're allowed to run to completion while the host is stopped. When the new instance gets the lease it will see that the timer is behind schedule and will start running the function again.

A related situation could happen if rather than simply losing the lease, the entire VM goes down while in the middle of the function invocation. Another instance will get the lease and start running that schedule occurrence (in this case which never completed). The first invocation is no longer running, but depending on at what point in its execution the VM went down it might have already sent some emails, etc. So you could do duplicate work in this case as well, even though the functions aren’t running concurrently. The general guidance is for people to write **idempotent** code that can handle retries (e.g. in queue processing where the first invocation might fail part way through), which is what this would effectively be at that point.
 
Our TimerTrigger listener lease ensures that the timer scheduler is only running on a **single app instance**. However, it isn’t enforcing a **single invocation** of the function, in error cases like the above it seems. I need to give this more thought. One idea for us to ensure that only a single instance of a timer function is running would be to bring the process down when the lease is lost, which would kill any running functions. Note that when using TimerTrigger in WebJobs classic a lost lease will cause the process to go down - so this is only an issue in functions because we're handling host errors and keeping the process alive.
 
**Workaround**: Schedule monitoring can be disabled by setting “useMonitor: false” in the function.json for your timer function. See [here](https://github.com/Azure/azure-webjobs-sdk-extensions/blob/45290c428ef168e2043fcc44bf4d73a30b866a6a/src/WebJobs.Extensions/Extensions/Timers/TimerTriggerAttribute.cs#L59) for details. It will no longer check past due on startup, which will prevent concurrent executions.

**Comment by ahmelsayed** [comment]
@mathewc should I close this as a dupe https://github.com/Azure/azure-functions-host/issues/2950?

**Comment by paulbatum** [comment]
I did it the other way round since this issue has the detailed explanation and workarounds.

**Comment by ahmelsayed** [comment]
yeah, sorry that's what I meant. Thanks Paul

**Comment by tomasopye** [comment]
Is there any sort of planned changes for this issue? Or are the workarounds all we get?

**Comment by mathewc** [comment]
Since this issue was originally logged, we've made improvements in this area. For example, in the latest versions of Azure Functions, when Timer functions loose their singleton leases for whatever reason, we'll shut the host process down, ensuring we terminate any in process invocation of the timer function. That, coupled with the guidance above on taking a cancellation token in your function to identify when the host is shutting down will minimize the possibility of concurrent invocations.

In summary, the platform improvements we've made over the years, coupled with the guidance above alleviates these issues.

**Comment by tomasopye** [comment]
@mathewc thank you very much for the response, we did notice where in appinsights a cancellation request was issued for one of the 2 running timer functions. I appreciate the guidance and we will be implementing the cancellation token in our timer triggers
